### DISCUSSION TITLE
Is there interest in adding abstract syntax tree tooling to graphql-java?

### DISCUSSION THREAD

**Discussion by chriswr95** [graphql-java/graphql-java#2615] [Ideas]
Hello from Twitter!

I've been hard at work for resident task master @jbellenger building a few pieces of functionality on top of the graphql-java AST:
* Get recursively referenced fragments: Given a node in the AST, get all fragments referenced by the node.
* Detect deprecated fields: Fairly straightforwards, detect if a query references a field marked deprecated in the schema.

Is there interest in bringing these pieces of functionality into graphql-java? Potentially via a tooling class that lives on the public api. For example: [ValidationContext](https://github.com/graphql/graphql-js/blob/81ca7781ec15e2d6b91d6146a3e545afb31af92b/src/validation/ValidationContext.ts) in graphql-js , and [DocumentAnalyzer](https://github.com/sangria-graphql/sangria/blob/c3900dbe8218d370b976ec7eac900616d7b53c8f/modules/core/src/main/scala/sangria/validation/DocumentAnalyzer.scala) in Sangria contain similar pieces of functionality.

If there is interest in seeing what this would look like, I can raise a pull request with the details.

**Comment by bbakerman** [comment]
There is interest.  It would be great to have a little design doc of what this might be say.

We have much of the machinery for this already via Ast visitor classes

```
        AstTransformer astTransformer = new AstTransformer();
        Node newDoc = astTransformer.transform(nodeToBeSorted, visitor);
```

(a transformer that does not change anything is a visitor)

and there is also a Query traverser that allows easy access to find things like deprecated fields or ones with special directives (we have a Beta directive in our graph at Atlassian)

```
        QueryTraverser queryTraverser = QueryTraverser.newQueryTraverser()
                .document(document)
                .operationName(operation)
                .schema(graphQLSchema)
                .variables(variables)
                .build();

        final List<String> deprecatedFieldList = new ArrayList<>();
        queryTraverser.visitPreOrder(new QueryVisitorStub() {
            @Override
            public TraversalControl visitFieldWithControl(QueryVisitorFieldEnvironment fieldEnvironment) {
                collectDeprecated(fieldEnvironment, deprecatedFieldList);
                return TraversalControl.CONTINUE;
            }
        });
        return deprecatedFieldList;

```

However know that there is other work going on in what we call `graphql.normalized.ExecutableNormalizedField`  - this represents the AST collapsed into its eventual "field" with schema types and so on all resolved

It turns out that many people think they want access to Fragments and so on to collect information BUT often they do not know enough about the way AST fields need to be merged to get it correct.  We have got it wrong over the years.

https://www.graphql-java.com/blog/deep-dive-merged-fields

So in short yes - I think some well defined helpers for doing specific things might be useful - but please make sure its implemented in terms of the generic visitors / traversers we already have

**Reply by chriswr95** [reply]
Thanks! I appreciate the detailed feedback. I've written a short design doc as a [separate discussion](https://github.com/graphql-java/graphql-java/discussions/2627). Please let me know your thoughts, and what I am missing
